---
output:
  md_document:
    variant: markdown_github
---
# Variant filtering  

## Tools & manuals  
vctools

## Methodology  

### Remove indels (insertions and deletions)  
Remove indels using *vcftools* v0.1.16 function ```--remove-indels```, leaving only single nucleotide polymorphisms (SNPs).   
**Code**  
```bash
vcftools --gzvcf doris29_variants.vcf.gz --remove-indels --out $OUTPUT_DIR/rm_indels.vcf --recode --recode-INFO-all
```   
**Data loss*
After filtering, kept 120429 out of a possible 123612 Sites.  

### Filter by missingness per sample  
No samples needed to be removed on the basis of missing data. All samples had low missing data, with the maximum missingness observed to be 5.35%.  
```{r, eval=TRUE}
setwd("G:/My Drive/Honours/Bioinformatics")
imiss <- read.table("missing_indv2.imiss", header = TRUE)

hist(imiss$F_MISS,
     breaks = 50,              # Number of bins
     main ="",
     xlab = "Proportion of missing variants (F_MISS)",
     ylab = "Number of individuals")

```  
### Filter by base quality  
Only sites with a minimum base quality score of 30 were retained, correlating 

```bash
vcftools --gzvcf doris29_raw.vcf.gz --minQ 30 --out $OUTPUT_DIR/rm_indels.vcf --recode --recode-INFO-all
```
After filtering, kept 113042 out of a possible 120429 Sites.   

Calculate mean depth and then filter by depth.
```bash
vcftools --vcf minQ30.recode.vcf --site-mean-depth --out $OUTPUT_DIR/meanDP

awk '{ total += $3 } END { print "Mean depth:", total/NR }' "$OUTPUT_DIR/meanDP.ldepth.mean"
```   
Filter by depth: 16 to 97 
```bash
module load vcftools
vcftools --vcf minQ30.recode.vcf --min-meanDP 16 --max-meanDP 97 --out $OUTPUT_DIR/site_depth --recode --recode-INFO-all
```   
After filtering, kept 78505 out of a possible 113042 Sites.  

Keep biallelic sites
```bash
module load vcftools
vcftools --vcf site_depth.recode.vcf --min-alleles 2 --max-alleles 2 --out $OUTPUT_DIR/biallelic_sites --recode --recode-INFO-all
```  
After filtering, kept 73993 out of a possible 78505 Sites.  


