---
output:
  github_document:
knitr:
  opts_chunk:
    fig.path: "figures/"
---
# Kinship Analysis

## Tools & manuals  
*vcftools* v0.1.16 [manual.](https://vcftools.github.io/man_latest.html)  

## Kinship dataset

### Removal of unrelated individuals  
Individuals that were found to genetically differ from the majority in the Palmer Archipelago were excluded from the kinship dataset. This was determined using structure and a PCA.  

Firstly, a list of individuals to exclude from the entire dataset was created, and saved as a text file: ```KINSHIP_remove.txt```.   
```
Individuals removed: 
RN02_L0107
RN02_L0108
RN02_L0120
RN02_L0124
RN02_L0131
RN02_L0137
RN02_L0138
RN02_L0147
RN02_L0153
RN02_L0126
RN02_L0130
RN02_L0132
RN02_L0133
RN02_L0134
RN02_L0135
RN02_L0136
RN02_L0125
RN02_L0127
RN02_L0146
RN02_L0149
RN02_L0150
RN02_L0152
RN02_L0154
RN02_L0156
RN02_L0157
RN02_L0032
RN02_L0033
RN02_L0035
RN02_L0039
RN02_L0040
RN02_L0041
RN02_L0043
RN02_L0045
RN02_L0047
RN02_L0050
RN02_L0053
RN02_L0054
RN02_L0060
RN02_L0061
RN02_L0163
RN02_L0164
RN02_L0067
RN02_L0056
RN02_L0062
RN02_L0139
RN02_L0141
RN02_L0142
RN02_L0145
RN02_L0148
RN02_L0158
RN02_L0160
```
```
vcftools --vcf site_depth.recode.vcf --remove KINSHIP_remove.txt --recode --recode-INFO-all --out KINSHIP_initial
```

As this kinship dataset comprised a new configuration of individuals, population specific SNP filtering was applied.  


After filtering, kept 73736 out of a possible 73736 Sites.

### Remove sites by missingness  

Site missingness was calculated using vcftools v0.1.16 functions --keep and --missing-site.

```
vcftools --vcf KINSHIP_initial.recode.vcf --max-missing 1.0 --recode --recode-INFO-all --out KINSHIP_maxmissing
```  

After filtering, kept 72236 out of a possible 73736 Sites.

### Filter by maximum observed heterozygosity  
Heterozygosity for a biallelic SNP should theoretically be ≤0.5. Values > 0.5 may represent paralogous loci or multilocus contigs (ref: https://onlinelibrary.wiley.com/doi/full/10.1111/j.1755-0998.2010.02967.x). Therefore, sites with an observed heterozygosity >0.5 were found and removed. This was done in R, using the packages ```adegenet``` and ```vcfr```. 

```
Required R packages:  
#install.packages("adegenet")
#install.packages("vcfR")
library(adegenet)
library(vcfR)

#calculating site heterozygosity for fine scale dispersal data  

#import vcf file into R
vcf_kinship <- read.vcfR("KINSHIP_CR1.recode.vcf")
#convert vcf data into a genind object
kinship_genind <- vcfR2genind(vcf_kinship)
#summary statistics
summary_kinship = summary(kinship_genind)
summary_kinship$Hobs
#export site heterozygosity into a csv file
write.csv(summary_kinship$Hobs, file = "kinship_Hobs_per_SNP.csv")
```  

Sites with a heterozygosity ≥ 0.7 can now be removed.  

```
vcftools --vcf KINSHIP_maxmissing.recode.vcf --exclude-positions Het_remove.txt --recode --recode-INFO-all --out KINSHIP_max_het
```

After filtering, kept 64089 out of a possible 72236 Sites.

### Filter by MAF  
```
vcftools --vcf KINSHIP_max_het.recode.vcf --maf 0.05 --recode --recode-INFO-all --out KINSHIP_maf0.05
```  
After filtering, kept 11576 out of a possible 64089 Sites.

### Select one SNP per locus
```
vcftools --vcf KINSHIP_maf0.05.recode.vcf --thin 500 --recode --recode-INFO-all --out KINSHIP_SNPs
```  
After filtering, kept 4341 out of a possible 11576 Sites.

##Kinship Analysis  

###Colony

###Sequoia

First the VCF file will be converted into a .RAW file that can be interpreted by Sequoia. 

```
#convert vcf file to plink
#plink --vcf KINSHIP_SNPs.recode.vcf --allow-extra-chr --recode --out KINSHIP_SNPs
#extract data from plink file
#plink --file KINSHIP_SNPs --allow-extra-chr --recodeA --out sequoia_KINSHIP_SNPs
```  

Sequoia was then run 

###MCMC-ped

I need to transfrom my .vcf file into .tped format. This can be done with plink.

```
plink --vcf KINSHIP_SNPs.recode.vcf --allow-extra-chr --make-bed --out MCMC_KINSHIP_BINARY
plink --bfile MCMC_KINSHIP_BINARY --allow-extra-chr --recode transpose --out MCMC_KINSHIP
```

Recode column 3 and 4 to 0 as genetic distance and physical distance is both unknown for ddRADseq data.

```
awk '{$3 = 0; $4 = 0; OFS=" "; print}' input.tped > output.tped
```
